File: PWA/scraper.py
"""
Stalkerware IoC Web Scraper — SQLite version
============================================
Source: https://github.com/AssoEchap/stalkerware-indicators

Builds a local SQLite database at public/stalkerware_iocs.db so the PWA
can fetch it as a static asset. No .env, no server, no credentials.

Dependencies: requests, pyyaml
    pip install requests pyyaml
"""

import os
import sys
import csv
import io
import sqlite3
import requests
import yaml
from datetime import date

# ─── Output location ──────────────────────────────────────────────────────
# Place the DB where the PWA expects it (public/ is copied as-is by Vite).
HERE = os.path.dirname(os.path.abspath(__file__))
DB_PATH     = os.path.join(HERE, "public", "stalkerware_iocs.db")
SCHEMA_PATH = os.path.join(HERE, "schema.sql")

# ─── Source URLs (ECHAP GitHub via jsDelivr CDN mirror) ───────────────────
SOURCES = {
    "ioc.yaml": [
        "https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/ioc.yaml",
        "https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/ioc.yaml",
    ],
    "watchware.yaml": [
        "https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/watchware.yaml",
        "https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/watchware.yaml",
    ],
    "samples.csv": [
        "https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/samples.csv",
        "https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/samples.csv",
    ],
}

SOURCE_REFERENCE = "ECHAP-AssoEchap-GitHub"

IOC_TYPE_MAP = {
    "domains":      "domain",
    "packages":     "package_name",
    "certificates": "certificate",
    "sha256":       "file_hash",
    "filenames":    "filename",
    "processes":    "process_name",
    "emails":       "email",
    "ips":          "ip_address",
}

SEVERITY_MAP = {
    "certificate":  "High",
    "package_name": "High",
    "file_hash":    "High",
    "ip_address":   "High",
    "domain":       "Medium",
    "process_name": "Medium",
    "filename":     "Medium",
    "email":        "Low",
}

RELATIONSHIP_MAP = {
    "domain":       "communicates_with",
    "ip_address":   "connects_to",
    "package_name": "uses",
    "file_hash":    "drops",
    "certificate":  "signed_with",
    "filename":     "drops",
    "process_name": "runs",
    "email":        "registered_with",
}

# ─── Fetch ────────────────────────────────────────────────────────────────

def fetch_url(filename):
    urls = SOURCES[filename]
    headers = {"User-Agent": "Mozilla/5.0 (stalkerware-ioc-scraper)"}
    for url in urls:
        try:
            print(f"    Trying {url} ...")
            resp = requests.get(url, headers=headers, timeout=15)
            resp.raise_for_status()
            print(f"    OK ({len(resp.content)} bytes)")
            return resp.text
        except Exception as e:
            print(f"    Failed: {e}")
    raise RuntimeError(f"Could not fetch {filename} from any source.")


# ─── DB helpers ───────────────────────────────────────────────────────────

def get_connection():
    os.makedirs(os.path.dirname(DB_PATH), exist_ok=True)
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")
    return conn


def get_or_insert_app(cur, app_name, platform, release_date):
    cur.execute("SELECT app_id FROM stalkerware_apps WHERE app_name = ?", (app_name,))
    row = cur.fetchone()
    if row:
        return row["app_id"]
    cur.execute(
        "INSERT INTO stalkerware_apps (app_name, platform, release_date) VALUES (?, ?, ?)",
        (app_name, platform, release_date.isoformat() if hasattr(release_date, "isoformat") else release_date),
    )
    return cur.lastrowid


def get_or_insert_ioc(cur, ioc_type, ioc_value, severity_level, date_identified):
    cur.execute(
        "SELECT ioc_id FROM iocs WHERE ioc_type = ? AND ioc_value = ?",
        (ioc_type, ioc_value),
    )
    row = cur.fetchone()
    if row:
        return row["ioc_id"]
    cur.execute(
        "INSERT INTO iocs (ioc_type, ioc_value, severity_level, date_identified) VALUES (?, ?, ?, ?)",
        (ioc_type, ioc_value, severity_level,
         date_identified.isoformat() if hasattr(date_identified, "isoformat") else date_identified),
    )
    return cur.lastrowid


def link_app_ioc(cur, app_id, ioc_id, relationship_type, source_reference):
    cur.execute(
        "SELECT app_ioc_id FROM app_ioc WHERE app_id = ? AND ioc_id = ?",
        (app_id, ioc_id),
    )
    if cur.fetchone():
        return
    cur.execute(
        "INSERT INTO app_ioc (relationship_type, source_reference, app_id, ioc_id) VALUES (?, ?, ?, ?)",
        (relationship_type, source_reference, app_id, ioc_id),
    )


# ─── Ingest YAML ──────────────────────────────────────────────────────────

def ingest_yaml(cur, filename, default_platform="Android"):
    print(f"\n  Fetching {filename} ...")
    raw = fetch_url(filename)
    entries = yaml.safe_load(raw)

    if not entries or not isinstance(entries, list):
        print(f"  [WARN] {filename} is empty or unexpected format.")
        return {"apps": 0, "iocs": 0, "links": 0}

    stats = {"apps": 0, "iocs": 0, "links": 0}
    today = date.today()

    for entry in entries:
        if not isinstance(entry, dict):
            continue

        name = str(entry.get("name", "")).strip()
        if not name:
            continue

        has_packages = bool(entry.get("packages"))
        platform = "Android" if has_packages else default_platform

        app_id = get_or_insert_app(cur, name, platform, today)
        stats["apps"] += 1

        for yaml_key, ioc_type in IOC_TYPE_MAP.items():
            values = entry.get(yaml_key, [])
            if not values:
                continue
            if isinstance(values, str):
                values = [values]

            for value in values:
                value = str(value).strip()
                if not value:
                    continue

                severity     = SEVERITY_MAP.get(ioc_type, "Medium")
                relationship = RELATIONSHIP_MAP.get(ioc_type, "related_to")

                ioc_id = get_or_insert_ioc(cur, ioc_type, value, severity, today)
                stats["iocs"] += 1
                link_app_ioc(cur, app_id, ioc_id, relationship, SOURCE_REFERENCE)
                stats["links"] += 1

    return stats


# ─── Ingest samples.csv ───────────────────────────────────────────────────

def ingest_samples_csv(cur):
    print(f"\n  Fetching samples.csv ...")
    raw = fetch_url("samples.csv")

    reader = csv.DictReader(io.StringIO(raw))
    stats  = {"iocs": 0, "skipped": 0}
    today  = date.today()

    for row in reader:
        sha256   = row.get("sha256", "").strip()
        app_name = row.get("app_name", "").strip()

        if not sha256 or not app_name:
            stats["skipped"] += 1
            continue

        cur.execute("SELECT app_id FROM stalkerware_apps WHERE app_name = ?", (app_name,))
        app_row = cur.fetchone()
        if not app_row:
            stats["skipped"] += 1
            continue

        app_id = app_row["app_id"]
        ioc_id = get_or_insert_ioc(cur, "file_hash", sha256, "High", today)
        link_app_ioc(cur, app_id, ioc_id, "drops", SOURCE_REFERENCE)
        stats["iocs"] += 1

    return stats


# ─── Main ─────────────────────────────────────────────────────────────────

def main():
    print("=" * 60)
    print("  Stalkerware IoC Scraper — SQLite (ECHAP source)")
    print("=" * 60)

    print(f"\n[0] Opening database at {DB_PATH} ...")
    conn = get_connection()
    cur = conn.cursor()
    with open(SCHEMA_PATH, "r", encoding="utf-8") as f:
        cur.executescript(f.read())
    print(f"    Schema applied from {os.path.basename(SCHEMA_PATH)}.")

    try:
        print("\n[1/3] Ingesting stalkerware — ioc.yaml ...")
        stats_ioc = ingest_yaml(cur, "ioc.yaml", default_platform="Android")
        print(f"      Apps: {stats_ioc['apps']} | IoCs: {stats_ioc['iocs']} | Links: {stats_ioc['links']}")

        print("\n[2/3] Ingesting watchware — watchware.yaml ...")
        stats_watch = ingest_yaml(cur, "watchware.yaml", default_platform="Android")
        print(f"      Apps: {stats_watch['apps']} | IoCs: {stats_watch['iocs']} | Links: {stats_watch['links']}")

        print("\n[3/3] Ingesting file hashes — samples.csv ...")
        stats_samples = ingest_samples_csv(cur)
        print(f"      Hashes inserted: {stats_samples['iocs']} | Skipped: {stats_samples['skipped']}")

        conn.commit()

        total_apps  = stats_ioc["apps"] + stats_watch["apps"]
        total_iocs  = stats_ioc["iocs"] + stats_watch["iocs"] + stats_samples["iocs"]
        total_links = stats_ioc["links"] + stats_watch["links"] + stats_samples["iocs"]

        print("\n" + "=" * 60)
        print("  Scrape Complete!")
        print(f"  Total apps inserted : {total_apps}")
        print(f"  Total IoCs inserted : {total_iocs}")
        print(f"  Total links created : {total_links}")
        print(f"  Database written to : {DB_PATH}")
        print("=" * 60)

    except Exception as e:
        conn.rollback()
        print(f"\n[ERROR] {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

    finally:
        cur.close()
        conn.close()


if __name__ == "__main__":
    main()

File: PWA/src/scraper.js
// Browser-side scraper. Mirrors PWA/scraper.py but builds the SQLite DB
// directly in the browser via sql.js. No server, no Python.
//
// Sources, IoC type map, severity map, and relationship map are kept in
// sync with scraper.py — keep both files updated together.

import yaml from 'js-yaml'
import schemaSql from '../schema.sql?raw'

const SOURCES = {
  'ioc.yaml': [
    'https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/ioc.yaml',
    'https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/ioc.yaml',
  ],
  'watchware.yaml': [
    'https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/watchware.yaml',
    'https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/watchware.yaml',
  ],
  'samples.csv': [
    'https://raw.githubusercontent.com/AssoEchap/stalkerware-indicators/master/samples.csv',
    'https://cdn.jsdelivr.net/gh/AssoEchap/stalkerware-indicators@master/samples.csv',
  ],
}

const SOURCE_REFERENCE = 'ECHAP-AssoEchap-GitHub'

const IOC_TYPE_MAP = {
  domains:      'domain',
  packages:     'package_name',
  certificates: 'certificate',
  sha256:       'file_hash',
  filenames:    'filename',
  processes:    'process_name',
  emails:       'email',
  ips:          'ip_address',
}

const SEVERITY_MAP = {
  certificate:  'High',
  package_name: 'High',
  file_hash:    'High',
  ip_address:   'High',
  domain:       'Medium',
  process_name: 'Medium',
  filename:     'Medium',
  email:        'Low',
}

const RELATIONSHIP_MAP = {
  domain:       'communicates_with',
  ip_address:   'connects_to',
  package_name: 'uses',
  file_hash:    'drops',
  certificate:  'signed_with',
  filename:     'drops',
  process_name: 'runs',
  email:        'registered_with',
}

async function fetchOne(filename, onProgress) {
  const urls = SOURCES[filename]
  let lastErr
  for (const url of urls) {
    try {
      onProgress?.(`Fetching ${filename} from ${new URL(url).hostname}…`)
      const res = await fetch(url, { cache: 'no-store' })
      if (!res.ok) throw new Error(`HTTP ${res.status}`)
      return await res.text()
    } catch (e) {
      lastErr = e
    }
  }
  throw new Error(`Could not fetch ${filename}: ${lastErr?.message || 'unknown error'}`)
}

// Tiny CSV parser that handles quoted fields with commas / embedded quotes.
// Sufficient for samples.csv from the ECHAP repo; not RFC-complete.
function parseCsv(text) {
  const rows = []
  let row = []
  let field = ''
  let inQuotes = false
  for (let i = 0; i < text.length; i++) {
    const c = text[i]
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') { field += '"'; i++ }
        else inQuotes = false
      } else field += c
    } else {
      if (c === '"') inQuotes = true
      else if (c === ',') { row.push(field); field = '' }
      else if (c === '\n') { row.push(field); rows.push(row); row = []; field = '' }
      else if (c === '\r') { /* skip */ }
      else field += c
    }
  }
  if (field.length || row.length) { row.push(field); rows.push(row) }
  if (!rows.length) return []
  const header = rows[0].map(h => h.trim())
  return rows.slice(1)
    .filter(r => r.some(v => v && v.length))
    .map(r => Object.fromEntries(header.map((h, i) => [h, (r[i] ?? '').trim()])))
}

// Insert helpers that mirror the python ones. We hold a map of in-flight
// IDs in JS instead of round-tripping a SELECT before every insert — the
// unique indexes in schema.sql still protect us if we somehow double-add.
function makeInserter(db) {
  const apps = new Map()  // app_name -> app_id
  const iocs = new Map()  // type|value -> ioc_id
  const links = new Set() // app_id|ioc_id

  const insertApp  = db.prepare(
    'INSERT INTO stalkerware_apps (app_name, platform, release_date) VALUES (?, ?, ?)'
  )
  const insertIoc  = db.prepare(
    'INSERT INTO iocs (ioc_type, ioc_value, severity_level, date_identified) VALUES (?, ?, ?, ?)'
  )
  const insertLink = db.prepare(
    'INSERT INTO app_ioc (relationship_type, source_reference, app_id, ioc_id) VALUES (?, ?, ?, ?)'
  )

  function lastInsertId() {
    const r = db.exec('SELECT last_insert_rowid() AS id')
    return r[0].values[0][0]
  }

  return {
    upsertApp(name, platform, releaseDate) {
      if (apps.has(name)) return apps.get(name)
      insertApp.run([name, platform, releaseDate])
      const id = lastInsertId()
      apps.set(name, id)
      return id
    },
    upsertIoc(type, value, severity, date) {
      const k = type + '|' + value
      if (iocs.has(k)) return iocs.get(k)
      insertIoc.run([type, value, severity, date])
      const id = lastInsertId()
      iocs.set(k, id)
      return id
    },
    linkAppIoc(appId, iocId, relationship, source) {
      const k = appId + '|' + iocId
      if (links.has(k)) return false
      insertLink.run([relationship, source, appId, iocId])
      links.add(k)
      return true
    },
    hasApp(name) { return apps.has(name) },
    getAppId(name) { return apps.get(name) },
    finalize() {
      insertApp.free()
      insertIoc.free()
      insertLink.free()
    },
  }
}

function ingestYaml(rawText, ins, defaultPlatform = 'Android') {
  const entries = yaml.load(rawText)
  const stats = { apps: 0, iocs: 0, links: 0 }
  if (!Array.isArray(entries)) return stats
  const today = new Date().toISOString().slice(0, 10)

  for (const entry of entries) {
    if (!entry || typeof entry !== 'object') continue
    const name = String(entry.name ?? '').trim()
    if (!name) continue
    const platform = entry.packages ? 'Android' : defaultPlatform
    const appId = ins.upsertApp(name, platform, today)
    stats.apps++

    for (const [yamlKey, iocType] of Object.entries(IOC_TYPE_MAP)) {
      let values = entry[yamlKey]
      if (!values) continue
      if (typeof values === 'string') values = [values]
      if (!Array.isArray(values)) continue

      const severity = SEVERITY_MAP[iocType] || 'Medium'
      const relationship = RELATIONSHIP_MAP[iocType] || 'related_to'

      for (const raw of values) {
        const value = String(raw ?? '').trim()
        if (!value) continue
        const iocId = ins.upsertIoc(iocType, value, severity, today)
        stats.iocs++
        if (ins.linkAppIoc(appId, iocId, relationship, SOURCE_REFERENCE)) stats.links++
      }
    }
  }
  return stats
}

function ingestSamplesCsv(rawText, ins) {
  const today = new Date().toISOString().slice(0, 10)
  const stats = { iocs: 0, skipped: 0 }
  const rows = parseCsv(rawText)
  for (const row of rows) {
    const sha256 = (row.sha256 || '').trim()
    const appName = (row.app_name || '').trim()
    if (!sha256 || !appName) { stats.skipped++; continue }
    // Only attach hashes to apps already ingested by the YAML pass.
    // ins doesn't expose the app map directly; we attempt the lookup via
    // upsertApp's cache by re-calling with no insert path — instead we
    // skip if upsertApp would create a fresh row. Cheapest way: peek.
    if (!ins.hasApp(appName)) { stats.skipped++; continue }
    const appId = ins.getAppId(appName)
    const iocId = ins.upsertIoc('file_hash', sha256, 'High', today)
    if (ins.linkAppIoc(appId, iocId, 'drops', SOURCE_REFERENCE)) stats.iocs++
  }
  return stats
}

// Build a brand-new sql.js DB and ingest everything. Returns the
// serialized Uint8Array ready to persist.
export async function buildDatabase({ sqlJs, onProgress } = {}) {
  if (!sqlJs) throw new Error('buildDatabase: sqlJs (the initialized SQL module) is required')
  onProgress?.('Initializing database…')
  const db = new sqlJs.Database()
  db.exec(schemaSql)

  const ins = makeInserter(db)
  let totalApps = 0, totalIocs = 0, totalLinks = 0
  try {
    db.exec('BEGIN')

    onProgress?.('Downloading ioc.yaml…')
    const iocYaml = await fetchOne('ioc.yaml', onProgress)
    onProgress?.('Parsing stalkerware entries…')
    const s1 = ingestYaml(iocYaml, ins)
    totalApps += s1.apps; totalIocs += s1.iocs; totalLinks += s1.links

    onProgress?.('Downloading watchware.yaml…')
    const watchYaml = await fetchOne('watchware.yaml', onProgress)
    onProgress?.('Parsing watchware entries…')
    const s2 = ingestYaml(watchYaml, ins)
    totalApps += s2.apps; totalIocs += s2.iocs; totalLinks += s2.links

    onProgress?.('Downloading samples.csv…')
    const samples = await fetchOne('samples.csv', onProgress)
    onProgress?.('Linking sample hashes…')
    const s3 = ingestSamplesCsv(samples, ins)
    totalIocs += s3.iocs; totalLinks += s3.iocs

    db.exec('COMMIT')
  } catch (e) {
    try { db.exec('ROLLBACK') } catch {}
    ins.finalize()
    db.close()
    throw e
  }

  ins.finalize()
  const bytes = db.export()
  db.close()
  onProgress?.(`Done. ${totalApps} apps, ${totalIocs} IoCs, ${totalLinks} links.`)
  return { bytes, stats: { apps: totalApps, iocs: totalIocs, links: totalLinks } }
}
